File: 10284/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

int chess[8][8];

int moves[][8][2] = {
    { {-1, -1}, {-1,  0}, {-1,  1}, {0,  1},  {1,  1}, {1,  0}, { 1, -1}, { 0, -1} },
    { {-1, -2}, {-2, -1}, { 1, -2}, {2, -1},  {1,  2}, {2,  1}, {-2,  1}, {-1,  2} }
};


int valid(int x, int y){
    return x >= 0 && y >= 0 && x < 8 && y < 8;
}


void set_attack(int x, int y){
    if(chess[x][y] == 0)
        chess[x][y] = 1;
}


void move(int x, int y, int mx, int my,int self){
    if(!valid(x, y)) return;
    if(self || chess[x][y] < 2){
        if(!self) set_attack(x, y);
            move(x+mx, y+my, mx, my, 0);
    }
}


void king_knight(int x, int y, int type){
    int i;
    for(i = 0; i < 8; i++)
        if(valid(x + moves[type][i][0], y + moves[type][i][1]))
            set_attack(x + moves[type][i][0], y + moves[type][i][1]);
}


void rook(int x, int y){
    int i;
    for(i = 1; i < 8; i += 2)
        move(x, y, moves[0][i][0], moves[0][i][1], 1);
}


void bishop(int x, int y){
   int i;
    for(i = 0; i < 8; i += 2)
        move(x, y, moves[0][i][0], moves[0][i][1], 1);
}


void queen(int x, int y){
    rook(x, y);
    bishop(x, y);
}


void pawn(int x, int y, int type){
    if(type){
        if(valid(x-1, y-1)) set_attack(x-1, y-1);
        if(valid(x-1, y+1)) set_attack(x-1, y+1);
    }
    else{
        if(valid(x+1, y-1)) set_attack(x+1, y-1);
        if(valid(x+1, y+1)) set_attack(x+1, y+1);
    }
}


void print_result(){
    int i,j, db = 0;
    for(i = 0; i < 8; i++)
        for(j = 0; j < 8; j++)
            if(chess[i][j] == 0) db++;
    printf("%d\n", db);
}


void init_chess(){
    int i,j;
    for(i = 0; i < 8; i++)
        for(j = 0; j < 8; j++)
            chess[i][j] = 0;
}


void set_chess(char *s){
    char *token = strtok(s, "/");
    int i, j, k;
    for(i = 0; token; i++){
        for(k = j = 0; token[k]; k++){
            if(isdigit(token[k])) j += token[k] - '0';
            else chess[i][j++] = token[k];
        }
        token = strtok(NULL, "/");
    }
}


void solve_chess(){
    int i, j, ch;
    for(i = 0; i < 8; i++){
        for(j = 0; j < 8; j++){
            ch = chess[i][j];
            if(     ch == 'q' || ch == 'Q') queen (i, j);
            else if(ch == 'r' || ch == 'R') rook  (i, j);
            else if(ch == 'b' || ch == 'B') bishop(i, j);
            else if(ch == 'k' || ch == 'K') king_knight(i, j, 0);
            else if(ch == 'n' || ch == 'N') king_knight(i, j, 1);
            else if(ch == 'p')              pawn(i, j, 0);
            else if(ch == 'P')              pawn(i, j, 1);
        }
    }
}


int main()
{
    char s[40];
    while(scanf("%s", s) != EOF){
        init_chess();
        set_chess(s);
        solve_chess();
        print_result();
    }
    return 0;
}

File: 10284/Main.java

import java.util.Scanner;

class Chess{
    private int chess[][];

    public Chess(String s) {
        chess = new int[8][8];
        String[] row = s.split("/");
        for(int i = 0; i < row.length; i++){
            int j, k;
            for(k = j = 0; k < row[i].length(); k++){
                if(Character.isDigit(row[i].charAt(k))) 
                    j += row[i].charAt(k) - '0';
                else 
                    chess[i][j++] = row[i].charAt(k);
            }
        }
    }
    
    public void move(int x, int y, int mx, int my, boolean self){
        if(!valid(x, y)) return;
        if(self || getFieldValue(x, y) < 2){
            if(!self) setAttack(x, y);
                move(x+mx, y+my, mx, my, false);
        }
    }
    
    public boolean valid(int x, int y){
        return x >= 0 && y >= 0 && x < 8 && y < 8;
    }
    
    public void setAttack(int x, int y){
        if(valid(x, y) && chess[x][y] == 0)
            chess[x][y] = 1;
    }
    
    public int getFieldValue(int x, int y){
        return chess[x][y];
    }
    
    public int getResult(){
        int db = 0;
        for(int i = 0; i < 8; i++)
            for(int j = 0; j < 8; j++)
                if(chess[i][j] == 0) db++;
        return db;
    }
        
}

class ChessSolver{
    Chess game;
    
    private static final int[][][] MOVES = new int[][][]{
                    { {-1, -1}, {-1,  0}, {-1,  1}, {0,  1},  {1,  1}, {1,  0}, { 1, -1}, { 0, -1} },
                    { {-1, -2}, {-2, -1}, { 1, -2}, {2, -1},  {1,  2}, {2,  1}, {-2,  1}, {-1,  2} }
    };

    public ChessSolver(String s) {
        this.game = new Chess(s);
    }
    
    void kingKnight(int x, int y, int type){
        for(int i = 0; i < 8; i++)
            game.setAttack(x + MOVES[type][i][0], y + MOVES[type][i][1]);
    }

    void rook(int x, int y){
        for(int i = 1; i < 8; i += 2)
            game.move(x, y, MOVES[0][i][0], MOVES[0][i][1], true);
    }

    void bishop(int x, int y){
        for(int i = 0; i < 8; i += 2)
            game.move(x, y, MOVES[0][i][0], MOVES[0][i][1], true);
    }

    void queen(int x, int y){
        rook(x, y);
        bishop(x, y);
    }

    void pawn(int x, int y, boolean type){
        if(type){
            game.setAttack(x-1, y-1);
            game.setAttack(x-1, y+1);
        }
        else{
            game.setAttack(x+1, y-1);
            game.setAttack(x+1, y+1);
        }
    }
    
    public int solveChess(){
        for(int i = 0; i < 8; i++){
            for(int j = 0; j < 8; j++){
                int ch = game.getFieldValue(i, j);
                switch (ch) { 
                    case 'q': case 'Q': queen (i, j);   break;
                    case 'r': case 'R': rook  (i, j);   break;
                    case 'b': case 'B': bishop(i, j);   break;
                    case 'k': case 'K': kingKnight(i, j, 0);  break;
                    case 'n': case 'N': kingKnight(i, j, 1);  break;
                    case 'p': pawn(i, j, false);    break;
                    case 'P': pawn(i, j, true);     break;
                    default: break;
                }
            }
        }
        return game.getResult();
    }
}

public class Main {
    public static void main(String[] args) {
        String s;
        Scanner sc = new Scanner(System.in);
        while(sc.hasNextLine()){
            ChessSolver cs = new ChessSolver(sc.nextLine());
            System.out.println(cs.solveChess());
        }
    }
}
